src/main/kotlin/org/rust/lang/core/parser/RustParserUtil.kt

```diff
@@ -434,15 +434,28 @@ object RustParserUtil : GeneratedParserUtilBase() {
         return true
     }
 
-    // foo ! ();
-    // ^ this name
-    private fun lookupSimpleMacroName(b: PsiBuilder): String? {
-        return if (b.tokenType == IDENTIFIER) {
-            val nextTokenIsExcl = b.probe {
-                b.advanceLexer()
-                b.tokenType == EXCL
+    // qualifier::foo ! ();
+    //            ^ this name
+    @Suppress("SameParameterValue")
+    private fun getMacroName(b: PsiBuilder, nameTokenIndex: Int): String? {
+        require(nameTokenIndex < 0) {
+            "`getMacroName` assumes that path with macro name is already parsed and the name token is behind of current position"
+        }
+
+        var steps = 0
+        var meaningfulSteps = 0
+
+        while (meaningfulSteps > nameTokenIndex) {
+            steps--
+            val elementType = b.rawLookup(steps) ?: return null
+
+            if (!isWhitespaceOrComment(b, elementType)) {
+                meaningfulSteps--
             }
-            if (nextTokenIsExcl) b.tokenText else null
+        }
+
+        return if (b.rawLookup(steps) == IDENTIFIER) {
+            PsiBuilderUtil.rawTokenText(b, steps).toString()
```

Comment:
Hmm, we have an extension `PsiBuilder.rawLookupText()` that does the same as `PsiBuilderUtil.rawTokenText` :thinking:

Reply:
Let's use own `PsiBuilder.rawLookupText()` since it checks corner cases

Reply:
Also, we have the same extention in toml plugin so I will move it into `common` module